### ISSUE TITLE
PMem performance not in line with expectations when using bdevperf

### ISSUE THREAD

**Issue by USTCqzy** [spdk/spdk#2107]
# Bug report

PMem performance not in line with expectations when using bdevperf

## Expected Behavior

One PMem max read performance ~ 7~8GB/s

## Current Behavior

Easily get results around 20GB/s

![pmem_randread](https://user-images.githubusercontent.com/49378451/130004636-3b14ad15-a882-4157-9f87-30c6da5d95d5.png)

## Possible Solution

No idea.

## Steps to Reproduce

1. ./bdevperf -q "$io_depth" -t 300 -w randread -o 4096 -m "$thread" -c test.json -C

test.json
```json
{
    "subsystems": [
      {
        "subsystem": "bdev",
        "config": [
          {
            "method": "bdev_pmem_create_pool",
            "params": {
              "num_blocks": 129884,
              "block_size": 4096,
              "pmem_file": "/mnt/myPMem/pmem_pool"
            }
          },
          {
            "method": "bdev_pmem_create",
            "params": {
              "name": "PMem0",
              "pmem_file": "/mnt/myPMem/pmem_pool"
            }
          }
        ]
      }
    ]
  }
```
pmem
```bash
$ fdisk -l
Disk /dev/pmem0: 124.3 GiB, 133175443456 bytes, 32513536 sectors
Units: sectors of 1 * 4096 = 4096 bytes
Sector size (logical/physical): 4096 bytes / 4096 bytes
I/O size (minimum/optimal): 4096 bytes / 4096 bytes

$ mount -o dax /dev/pmem0 /mnt/myPMem
```

## Context (Environment including OS version, SPDK version, etc.)

SPDK 21.07.x
Ubuntu 20.04.2

**Comment by changpe1** [comment]
@USTCqzy can you post your whole hardware configurations ?

**Comment by dong-liuliu** [comment]
From you test.json,               "num_blocks": 129884,  "block_size": 4096, the total size of pmem bdev is only around 512MB.
I guess all IO hit into the cache facility inside PMEM drive.

**Comment by USTCqzy** [comment]
@dong-liuliu
As I change "num_blocks" to 12988400, performance is still over 20GB/s for some conditions.

**Comment by USTCqzy** [comment]
@changpe1 Here is some info about cpu and memory
$ cat /proc/cpuinfo | grep name | cut -f2 -d: | uniq -c
     80  Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
$ cat /proc/meminfo |head -20
MemTotal:       263568600 kB
MemFree:         3989512 kB
MemAvailable:   138479848 kB
Buffers:           94520 kB
Cached:         136473012 kB
SwapCached:         1360 kB
Active:          7184008 kB
Inactive:       131480872 kB
Active(anon):    3296820 kB
Inactive(anon):  1996668 kB
Active(file):    3887188 kB
Inactive(file): 129484204 kB
Unevictable:       18588 kB
Mlocked:           18588 kB
SwapTotal:       8388604 kB
SwapFree:        8106748 kB
Dirty:              1740 kB
Writeback:             0 kB
AnonPages:       2115752 kB
Mapped:           379932 kB

**Comment by USTCqzy** [comment]
I would like to add my new thought (Hope to discuss!): Instead of pmem bandwidth being too high, maybe iops is the core metrics, which is too low actually.
Here are some results:
When pmem pool block size & io size is 256B (aligned with Intel's report: [https://www.intel.com/content/www/us/en/products/docs/memory-storage/optane-persistent-memory/optane-persistent-memory-200-series-brief.html](url))
As the following figure showed, IOPS is no more than 6 million, much lower than ~30M showed in the report.
![Figure_5_pmem_256_randread_iops](https://user-images.githubusercontent.com/49378451/130316740-c1cb9f9e-8098-4f77-a260-d4fd38811c17.png)
When pmem pool block size & io size is 4KB, it still seems failing to reach the theoretical point (~30Million)
![Figure_5_pmem_4096_randread_iops](https://user-images.githubusercontent.com/49378451/130316744-5f37859d-62b3-45ea-9796-3278b336280c.png)

Now I am wondering whether SPDK can fully utilize PMem.

**Comment by Comphix** [comment]
@USTCqzy , could you share us more information about your expectation of using SPDK with PMem? And in which situation and workloads for your usage so that we could also better understand SPDK for PMem. Thanks, Gang.

**Comment by USTCqzy** [comment]
@Comphix 
Thanks for your advice.
I would like to use ocf to construct a cache, combining DRAM+PMem (in App Direct Mode), but I found ocf performance much worse than I expect. To figure out, I explore performance of DRAM and PMem separately, and confuse with the performance gap between what Intel's claimed (30M IOPS) and my test results (6M IOPS).
If there is something unclear in the implementation of pmem bdev which makes it not working well, I might reconsider of using ocf.

PS: In previous study, IOPS bound can reach in ~4 thd, which also seems not in line with SPDK results.
Reference:
![image](https://user-images.githubusercontent.com/49378451/130323254-23accb75-2bed-4dda-b6c5-2bd390896cc6.png)
[https://www.usenix.org/system/files/fast20-yang.pdf](url)

**Comment by Comphix** [comment]
From our understanding, PMem's high performance is CPU intensive and Sync operations (like you have done to use many cores) , so that no matter the queue depth, it will have similar performance (i.e., queue depth = 1). Just to make sure in the first diagram, it could achieve 24GB for 64 cores, is it also done by SPDK Bdevperf?

SPDK's one of the Key principles is Async operations when the backend device supports, so that SPDK could set a high queue depth and with less CPU cores to achieve higher performance.

**Comment by USTCqzy** [comment]
@Comphix 
"Just to make sure in the first diagram, it could achieve 24GB for 64 cores, is it also done by SPDK Bdevperf?"
To clarify, the diagrams without reference are drawed by myself, and such datas are from experiments using SPDK bdevperf (commands showed before: `./bdevperf -q "$io_depth" -t 300 -w randread -o 4096 -m "$thread" -c test.json -C`).
And compared with the results from Intel's report and fast20 paper, I found that bdevperf got much worse performance.

"SPDK's one of the Key principles is Async operations when the backend device supports, so that SPDK could set a high queue depth and with less CPU cores to achieve higher performance."
In my experiments, io depth makes no difference in pmem bdev situations, e.g.
![Figure_2_pmem_4096_randread](https://user-images.githubusercontent.com/49378451/130351591-229181a1-f3da-4ed7-a849-7a673929bc74.png)

In conclusion, I would like to figure out where performance loss comes from, and I am also curious about how to use pmem bdev more efficiently (e.g. "set a high queue depth and with less CPU cores to achieve higher performance")

**Comment by Comphix** [comment]
I am a little confused. 

So with SPDK bdevperf as below diagram (32 threads + queue depth = 1) could get 24GB/s bandwidth

![PMem](https://user-images.githubusercontent.com/5643608/130391917-cf611c8e-4c60-410d-b5b1-a18ed7705140.PNG)

The link below is for the second generation of PMem (which is PMem 200) and could you share us what's your PMem configuration?
https://www.intel.com/content/www/us/en/products/docs/memory-storage/optane-persistent-memory/optane-persistent-memory-200-series-brief.html

For the Paper, the performance is about 6GB for 25 threads and this also not matched with SPDK Bdevperf performance...

**Comment by USTCqzy** [comment]
@Comphix 
"The link below is for the second generation of PMem (which is PMem 200) and could you share us what's your PMem configuration?"
Our PMem configuration is Intel® Optane™ Persistent Memory 100 Series - 128GB
[19-040-intel-optane-dc-memory-pb-low.pdf](https://github.com/spdk/spdk/files/7029376/19-040-intel-optane-dc-memory-pb-low.pdf)
It is around 6.8GB/s for 100%read+256B IO and 1.7GB/s for 100%read+64B IO.
![image](https://user-images.githubusercontent.com/49378451/130396104-2bf6afa8-489f-481c-9357-7d6085df1563.png)

In my experiments, it seems IOPS bound is more important. I mean, when IO size grows, IOPS unchanges and bandwidth grows.
This diagram is testing in 256B IO size (**I expected to get 6.8GB/s**).
![Figure_2_pmem_256_randread](https://user-images.githubusercontent.com/49378451/130396347-8ca778ad-c18f-45eb-8683-f6f6f52dd298.png)
This diagram is testing in 4KB IO size.
![Figure_2_pmem_4096_randread](https://user-images.githubusercontent.com/49378451/130396353-65b42da8-c247-494b-84d9-96be14a3e300.png)

**Comment by Comphix** [comment]
For the 256B IO size for SPDK Bdevperf, one thing is that it is better to configure the PMem to 512B sector size, currently from the description, it is configured with 4096.

The next thing is that Block IO usually align (and multiple) for the sector size, so for 256B IO, it will turn into 512 (if 512 is configured for the sector size). You may have a try of 512B IO first after the PMem is configured with 512B sector size (i.e., block_size below).

{
    "subsystems": [
      {
        "subsystem": "bdev",
        "config": [
          {
            "method": "bdev_pmem_create_pool",
            "params": {
              "num_blocks": 129884,
              "block_size": 4096,
              "pmem_file": "/mnt/myPMem/pmem_pool"
            }
          },
          {
            "method": "bdev_pmem_create",
            "params": {
              "name": "PMem0",
              "pmem_file": "/mnt/myPMem/pmem_pool"
            }
          }
        ]
      }
    ]
  }

**Comment by USTCqzy** [comment]
@Comphix 
I am wondering whether bdevperf has flag to avoid influence from memory page cache. Thx!

**Comment by optimistyzy** [comment]
> @Comphix
> I am wondering whether bdevperf has flag to avoid influence from memory page cache. Thx!

If  you want to bypass the cache for pmem test formatted with fsdax mode. You can mount your file system with "-o dax" choice.

**Comment by jimharris** [comment]
This doesn't appear to be an SPDK-related issue.  Feel free to ask more questions on this issue thread though.
